core_bpe.go

```diff
@@ -21,12 +21,7 @@ type CoreBPE struct {
 	sortedTokenBytes     [][]byte
 }
 
-func NewCoreBPE(encoder map[string]int, specialTokensEncoder map[string]int, pattern string) (*CoreBPE, error) {
-	regex, err := regexp2.Compile(pattern, regexp2.None)
-	if err != nil {
-		return nil, fmt.Errorf("error compiling regex: %s", err)
-	}
-
+func NewCoreBPE(encoder map[string]int, specialTokensEncoder map[string]int, regex *regexp2.Regexp) (*CoreBPE, error) {
```

Comment:
The regex compilation needs to be known at [compile time](https://github.com/dlclark/regexp2cg?tab=readme-ov-file#notes), which means that we need to compile the regex higher up the stack, hence the reason why we pass regexp2.Regexp here instead of a string.